### PR TITLE
Make DenseLayout transpiler pass target aware

### PR DESCRIPTION
**Pull request by mtreinish**
### Summary

This commit updates the dense layout pass target aware so that at
its instantiation a Target object can be specified to define the target
device for compilation. When specified this target will be used for
all device aware operations in the pass. When a target is specified the
error matrix used is the average of all 1q and 2q operations while
before the matrix was composed of just cx and readout error rates.

### Details and comments

Part of #7113

### CODE REVIEW

File: qiskit/transpiler/passes/layout/dense_layout.py
Comment (mtreinish) [reply qiskit/transpiler/passes/layout/dense_layout.py:160]: Done in: https://github.com/Qiskit/qiskit-terra/pull/7776/commits/683120f9f5169b29adde6abd74043867cfaaf36c

File: qiskit/transpiler/passes/layout/dense_layout.py
Comment (itoko) [inline qiskit/transpiler/passes/layout/dense_layout.py:93]: It seems to change the logic, is it intended? The original code counts the number of ops in "circuit" while the new code sets the constants (~= the number of ops defined in "target"?).

Comment (mtreinish) [reply qiskit/transpiler/passes/layout/dense_layout.py:93]: Yeah it was an intentional chamge, see the commit message on: https://github.com/Qiskit/qiskit-terra/pull/7776/commits/4fe2c44c8915a623a1ac668ffabd892c8ebdb402 It's mainly because at the layout stage there isn't an easy way to determine the ratio of 1q to 2q operations in the circuit besides iterating over the entire circuit and counting (which adds a lot of overhead). Also so early in the pass manager I'm not sure how accurate scaling the error based on the number of gates will be because we haven't done routing, basis translation, or any optimization passes. Since it's just a heuristic I figured just using the raw error rates was sufficient.

But I definitely am not 100% sure on this, if you have a suggestion for a better way to do this  I'm definitely open to making this better. The 2q error rates weren't even working before #7740 (which fixed the bug) so we can take this opportunity to come up with a better heuristic for weighing the error of the layouts.

File: releasenotes/notes/dense-layout-target-aware-2b330ccee948d31a.yaml
Comment (kevinhartman) [inline releasenotes/notes/dense-layout-target-aware-2b330ccee948d31a.yaml:5]:
```suggestion
    on its constructor, ``target``. This argument is used to specify a
```
```suggestion
    on it's constructor, ``target``. This argument is used to specify a
```

File: qiskit/transpiler/passes/layout/dense_layout.py
Comment (kevinhartman) [inline qiskit/transpiler/passes/layout/dense_layout.py:38]: Is there a good way we can make `coupling_map` optional when `target` is provided? I'm guessing the answer is "no", but just curious.

Comment (kevinhartman) [inline qiskit/transpiler/passes/layout/dense_layout.py:38]: Is there a good way we can make `coupling_map` optional when `target` is provided? I'm guessing the answer is "no", but just curious.

**Review by kevinhartman — APPROVED** [review]
LGTM.

File: qiskit/transpiler/passes/layout/dense_layout.py
Comment (mtreinish) [reply qiskit/transpiler/passes/layout/dense_layout.py:38]: I can set a default on `coupling_map` to `None`. I did this other passes, not sure why I missed this here. The pass already supports a user explicitly setting `coupling_map=None` in the constructor (but it'll fail at runtime). You'll be required to set one or the other or it will fail at runtime, which isn't any different that manually setting `coupling_map` to `None` today

Comment (mtreinish) [reply qiskit/transpiler/passes/layout/dense_layout.py]: Done in: https://github.com/Qiskit/qiskit-terra/pull/7776/commits/5814d0249bdc88ac7b5675694f1ee273a0730d6c

File: qiskit/transpiler/passes/layout/dense_layout.py
Comment (kevinhartman) [inline qiskit/transpiler/passes/layout/dense_layout.py:78]: Would this ever not be set by the initializer? Just curious why this assertion is here rather than at the end of the init.

File: qiskit/transpiler/target.py
Comment (kevinhartman) [inline qiskit/transpiler/target.py:516]:
```suggestion
        # Cast a list if passed in by mistake
```

Maybe?

**Review by kevinhartman — APPROVED** [review]
LGTM.

File: qiskit/transpiler/passes/layout/dense_layout.py
Comment (mtreinish) [reply qiskit/transpiler/passes/layout/dense_layout.py:78]: It's a weird quirk with the layout and routing passes. The way they get initialized in the preset pass managers is if there is not connectivity constraint they get initialized with `coupling_map=None` (or now a target without qubits) and then the output pass manager doesn't include the pass. If we raised an error at init the function building the preset pass managers would fail blocking a `transpile()` even though not having a connectivity constraint is valid. It's something I want to cleanup after #6403 so we don't waste time constructing passes we're not using but at least for right now it's necessary to not raise during `__init__`

File: qiskit/transpiler/target.py
Comment (mtreinish) [reply qiskit/transpiler/target.py]: Yeah it should be that, but this was part of #7778 so we can't fix it as part of this PR (it's probably only showing up in the incremental diff because of the merge/rebase in https://github.com/Qiskit/qiskit-terra/pull/7776/commits/b4afbc01bc24dc1196e9cbd8df3e628910a8bdc2).

Comment (kevinhartman) [reply qiskit/transpiler/target.py]: Ah yeah that's exactly what happened.

**Review by kevinhartman — APPROVED** [review]

**Comment by jakelishman** [comment]
@kevinhartman: you've got the permissions (and the authority) to merge after you've given your tick.  You can tag things with the "automerge" label, and then the bot will automatically merge them if tests pass, and will handle trains of PRs (bring PR 1 up-to-date, merge, bring PR 2 up-to-date, merge, etc) for you/us.  That way you don't have to wait around, or remember to come back once the tests are done, which is super helpful.

(The bot only merges if all the statuses are good: tests are passed, branch is up-to-date, PR is approved by code owner, and the "on hold" label isn't applied.)

**Pull request merged**
